components/sequencer/mysql/mysql.go

```diff
@@ -0,0 +1,158 @@
+//
+// Copyright 2021 Layotto Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package mysql
+
+import (
+	"database/sql"
+	"fmt"
+	"mosn.io/layotto/components/pkg/utils"
+	"mosn.io/layotto/components/sequencer"
+	"mosn.io/pkg/log"
+)
+
+type MySQLSequencer struct {
+	metadata   utils.MySQLMetadata
+	biggerThan map[string]int64
+	logger     log.ErrorLogger
+	db         *sql.DB
+}
+
+func NewMySQLSequencer(logger log.ErrorLogger) *MySQLSequencer {
+	s := &MySQLSequencer{
+		logger: logger,
+	}
+
+	return s
+}
+
+func (e *MySQLSequencer) Init(config sequencer.Configuration) error {
+
+	m, err := utils.ParseMySQLMetadata(config.Properties)
+
+	if err != nil {
+		return err
+	}
+	e.metadata = m
+	e.metadata.Db = e.db
+	e.biggerThan = config.BiggerThan
+
+	if err = utils.NewMySQLClient(e.metadata); err != nil {
+		return err
+	}
+
+	if len(e.biggerThan) > 0 {
+		var Key string
+		var Value int64
+		for k, bt := range e.biggerThan {
+			if bt <= 0 {
+				continue
+			} else {
+				m.Db.QueryRow(fmt.Sprintf("SELECT sequencer_key,sequencer_value FROM %s WHERE sequencer_key = ?", m.TableName), k).Scan(&Key, &Value)
+				if Value < bt {
+					return fmt.Errorf("MySQL sequencer error: can not satisfy biggerThan guarantee.key: %s,key in MySQL: %s", k, Key)
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (e *MySQLSequencer) GetNextId(req *sequencer.GetNextIdRequest) (*sequencer.GetNextIdResponse, error) {
+
+	metadata, err := utils.ParseMySQLMetadata(req.Metadata)
+	metadata.Db = e.db
+	var Key string
+	var Value int64
+	if err != nil {
+		return nil, err
+	}
+
+	begin, err := metadata.Db.Begin()
+	if err != nil {
+		return nil, err
+	}
+
+	err = begin.QueryRow("SELECT sequencer_key,sequencer_value FROM ? WHERE sequencer_key = ?", metadata.TableName, req.Key).Scan(&Key, &Value)
+
+	if err == sql.ErrNoRows {
+		Value = 1
+		_, err := begin.Exec("INSERT INTO ?(sequencer_key, sequencer_value) VALUES(?,?)", metadata.TableName, req.Key, Value)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		Value += 1
+		_, err := begin.Exec("UPDATE ? SET sequencer_value += 1 WHERE sequencer_key = ?", metadata.TableName, req.Key)
```

Comment:
might occur concurrency problem , such as   clientA and clientB all want to add the sequencer form 1 to 2 , the db sequencer added to 3，but the both client will get 2.

Reply:
Do I need to use SnowFlake? Or simply add a lock to the method.

Reply:
global lock? not very good.  try  optimistic lock ？